### ISSUE TITLE
pyright faulty diagnostics for AnyStr = TypeVar('AnyStr', bytes, str) on generic class (Python os)

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#2398]
This is a tracking bug. See https://github.com/microsoft/pylance-release/issues/1917 for original.

This appears to be specific to a context manager use case.

```python
import os
from typing import AnyStr

def thing(value: AnyStr):
    with os.scandir(value) as it:
        for file in it:
            if file.name.endswith('.xml'): ...
```

**Comment by shoffmeister** [comment]
FWIW, there are actually two problems for pyright in the above code snippet:
* https://github.com/microsoft/pylance-release/issues/1917 is about TypeVar with enumerated type instance constraints (anyone with a better term?); this is the real issue tracked here I think (i.e. without `def thing(value: AnyStr):`)
* https://github.com/microsoft/pylance-release/issues/1918 which I reported as a separate problem and which Eric kindly added as https://github.com/microsoft/pyright/issues/2399

IOW: remove `def thing(value: AnyStr):` which triggers https://github.com/microsoft/pyright/issues/2399

/me loves the single source of truth concept ;)

**Comment by erictraut** [comment]
This problem is arguably a bug in the typeshed stub `os/__init__.pyi`. It defines a class called `_ScandirIterator` that uses a cyclical definition that should probably be considered an error.

```python
class _ScandirIterator(Iterator[DirEntry[AnyStr]], ContextManager[_ScandirIterator[AnyStr]]): ...
```

The problem is that the arity of the generic type (i.e. the number of type arguments it accepts) is defined implicitly by the number of unique type variables that appear within the class declaration. A type checker needs to determine the arity before it can validate whether a specialization of that class is valid (provides type arguments that match that arity and that are compatible with the specified type parameters). In this case, however, the subexpression `_ScandirIterator[AnyStr]` needs to be validated before the arity of the generic type is established.

Consider, for example, if the class declaration were changed slightly to the following:
```python
class _ScandirIterator(Iterator[DirEntry[AnyStr]], ContextManager[_ScandirIterator[T]]): ...
```

Now its arity is 2 because `AnyStr` and `T` are both type variables, and `_ScandirIterator[T]` is now invalid because it provides only one type argument.

I had code in place within pyright to detect and report this type of ambiguous cyclical dependency, but I had to remove it because these constructs do appear in some stubs and mypy appears to ignore the problem.

I've updated the logic in pyright to deal better with this particular case, so the use of `os.scandir` works properly.
